crates/ty_python_semantic/src/types.rs

```diff
@@ -4962,20 +4962,16 @@ impl<'db> Type<'db> {
                 Type::FunctionLiteral(function.apply_specialization(db, specialization))
             }
 
-            // Note that we don't need to apply the specialization to `self_instance`, since it
-            // must either be a non-generic class literal (which cannot have any typevars to
-            // specialize) or a generic alias (which has already been fully specialized). For a
-            // generic alias, the specialization being applied here must be for some _other_
-            // generic context nested within the generic alias's class literal, which the generic
-            // alias's context cannot refer to. (The _method_ does need to be specialized, since it
-            // might be a nested generic method, whose generic context is what is now being
-            // specialized.)
             Type::BoundMethod(method) => Type::BoundMethod(BoundMethodType::new(
                 db,
                 method.function(db).apply_specialization(db, specialization),
-                method.self_instance(db),
+                method.self_instance(db).apply_specialization(db, specialization),
             )),
 
+            Type::NominalInstance(instance) => Type::NominalInstance(
+                instance.apply_specialization(db, specialization),
+            ),
```

Comment:
we probably need to do the same for the `Type::ProtocolInstance` variant, since protocols can also be generic. But I can do that as part of https://github.com/astral-sh/ruff/pull/17832, if you like, since that's the "hook up the generics infrastructure with the protocols infrastructure" PR 😄

Reply:
I did this in https://github.com/astral-sh/ruff/pull/17832/commits/e21d81b8ca17e02f207727c6f5e7d560c2456f8d as part of #17832